### DISCUSSION TITLE
Enhancement: Compile calc under Windows

### DISCUSSION THREAD

**Discussion by Leoongithub** [lcn2/calc#66] [Ideas]
All I can find about compiling calc under Windows is many years ago. Although calc is available under Windows via WSL, I think an exe file is still useful. I've also tried compiling calc without Cygwin, MSYS, etc., but it never worked. Here are some notes about this.

In short, the easiest way to compile calc is using Cygwin as stated in README.WINDOWS. But use `make all target=simple` instead of `make all target=Cygwin`, or directly use Makefile.simple instead of Makefile as HOWTO.INSTALL said. When using the "target=Cygwin" parameter, I get the following error.

```
x86_64-w64-mingw32-gcc  -DCALC_SRC -DCUSTOM -Wall -Wextra -pedantic     -UUNBAN  -O3 -g3   -c -o comfunc.o comfunc.c
comfunc.c:42:27: error: initializer element is not constant
   42 | STATIC NUMBER _q10_ = { { _tenval_, 1, 0 }, { _oneval_, 1, 0 }, 1, NULL };
      |                           ^~~~~~~~
comfunc.c:42:27: note: (near initialization for '_q10_.num.v')
comfunc.c:42:47: error: initializer element is not constant
   42 | STATIC NUMBER _q10_ = { { _tenval_, 1, 0 }, { _oneval_, 1, 0 }, 1, NULL };
      |                                               ^~~~~~~~
comfunc.c:42:47: note: (near initialization for '_q10_.den.v')
comfunc.c:43:26: error: initializer element is not constant
   43 | STATIC NUMBER _q0_ = { { _zeroval_, 1, 0 }, { _oneval_, 1, 0 }, 1, NULL };
      |                          ^~~~~~~~~
comfunc.c:43:26: note: (near initialization for '_q0_.num.v')
comfunc.c:43:47: error: initializer element is not constant
   43 | STATIC NUMBER _q0_ = { { _zeroval_, 1, 0 }, { _oneval_, 1, 0 }, 1, NULL };
      |                                               ^~~~~~~~
```

I've tried editing this file to remove the error, but there are still other files that don't compile. If you use the `target=simple` parameter, Cygwin will use `cc` instead of `x86_64-w64-mingw32-gcc`. I'm not sure about difference between these. Here is information for these two programs.

```
$ x86_64-w64-mingw32-gcc -v
Using built-in specs.
COLLECT_GCC=x86_64-w64-mingw32-gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-w64-mingw32/11/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: /mnt/share/cygpkgs/mingw64-x86_64-gcc/mingw64-x86_64-gcc.x86_64/src/gcc-11.3.0/configure --srcdir=/mnt/share/cygpkgs/mingw64-x86_64-gcc/mingw64-x86_64-gcc.x86_64/src/gcc-11.3.0 --prefix=/usr --exec-prefix=/usr --localstatedir=/var --sysconfdir=/etc --docdir=/usr/share/doc/mingw64-x86_64-gcc --htmldir=/usr/share/doc/mingw64-x86_64-gcc/html -C --build=x86_64-pc-cygwin --host=x86_64-pc-cygwin --target=x86_64-w64-mingw32 --without-libiconv-prefix --without-libintl-prefix --with-sysroot=/usr/x86_64-w64-mingw32/sys-root --with-build-sysroot=/usr/x86_64-w64-mingw32/sys-root --disable-multilib --disable-win32-registry --enable-languages=c,c++,fortran,lto,objc,obj-c++ --enable-fully-dynamic-string --enable-graphite --enable-libgomp --enable-libquadmath --enable-libquadmath-support --enable-libssp --enable-version-specific-runtime-libs --enable-libgomp --enable-libada --with-dwarf2 --with-gcc-major-version-only --with-gnu-ld --with-gnu-as --with-tune=generic --with-cloog-include=/usr/include/cloog-isl --with-system-zlib --enable-threads=posix --libexecdir=/usr/lib
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.3.0 (GCC)
```

```
$cc -v
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-cygwin/11/lto-wrapper.exe
Target: x86_64-pc-cygwin
Configured with: /mnt/share/cygpkgs/gcc/gcc.x86_64/src/gcc-11.3.0/configure --srcdir=/mnt/share/cygpkgs/gcc/gcc.x86_64/src/gcc-11.3.0 --prefix=/usr --exec-prefix=/usr --localstatedir=/var --sysconfdir=/etc --docdir=/usr/share/doc/gcc --htmldir=/usr/share/doc/gcc/html -C --build=x86_64-pc-cygwin --host=x86_64-pc-cygwin --target=x86_64-pc-cygwin --without-libiconv-prefix --without-libintl-prefix --libexecdir=/usr/lib --with-gcc-major-version-only --enable-shared --enable-shared-libgcc --enable-static --enable-version-specific-runtime-libs --enable-bootstrap --enable-__cxa_atexit --with-dwarf2 --with-tune=generic --enable-languages=c,c++,fortran,lto,objc,obj-c++,jit --enable-graphite --enable-threads=posix --enable-libatomic --enable-libgomp --enable-libquadmath --enable-libquadmath-support --disable-libssp --enable-libada --disable-symvers --with-gnu-ld --with-gnu-as --with-cloog-include=/usr/include/cloog-isl --without-libiconv-prefix --without-libintl-prefix --with-system-zlib --enable-linker-build-id --with-default-libstdcxx-abi=gcc4-compatible --enable-libstdcxx-filesystem-ts
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.3.0 (GCC)
```

I'd love to try some other ways to build calc under Windows if anyone can provide relevant information and help.

For those who want to use calc under Windows but don't have the environment, you can try to download the executable file I compiled. This zip file includes the calc executable program, the dll files needed to run it, headers, libraries, help files and some scripts. It should be noted that the help function cannot be used outside the Cygwin environment. If you just want calc itself, there are only 4 necessary files: calc.exe, cygwin1.dll, cygreadline7.dll, cygncursesw-10.dll.

[calc-2.14.1.2.zip](https://github.com/lcn2/calc/files/10152163/calc-2.12.7.2.zip)

2023.2.15 update:
[calc-2.14.1.3.zip](https://github.com/lcn2/calc/files/10740538/calc-2.14.1.3.zip)

2023.3.21 update:
[calc-2.14.1.5.zip](https://github.com/lcn2/calc/files/11025458/calc-2.14.1.5.zip)

2023.7.4 update:
[calc-2.14.1.6.zip](https://github.com/lcn2/calc/files/11946448/calc-2.14.1.6.zip)

2023.8.13 update:
[calc-2.14.3.0.zip](https://github.com/lcn2/calc/files/12328806/calc-2.14.3.0.zip)

**Comment by lcn2** [comment]
Hello @Leoongithub,

Thank you for your bug report.  While the maintainers of calc do not have a Window's license, we are working with someone within Microsoft (on occasion when their scheduled permits) to try and modify calc to be able to be compiled under Windows 11 using Microsoft's publicly available development tools.

We will look at what you have provided above as well.

If we can add the proper `#define` and `#if` stuff that allows a Windows 11 user to compile and use calc .. assume that it does not require overcomplicating calc code AND of course assuming it does NOT impact non-Windows users on any way, we will be happy to do so.  We will update `README.WINDOWS` accordingly.

Again, thanks for your input.  We will keep this open and, if you don't mind, we would add you to try things from time to time.

We will also give consideration to your idea of installing a _zip file for windows calc_ as well.

**Comment by lcn2** [comment]
> ...
> 
> I'd love to try some other ways to build calc under Windows if anyone can provide relevant information and help.
> 
> For those who want to use calc under Windows but don't have the environment, you can try to download the executable file I compiled. This zip file includes the calc executable program, the dll files needed to run it, headers, libraries, help files and some scripts. It should be noted that the help function cannot be used outside the Cygwin environment. If you just want calc itself, there are only 4 necessary files: calc.exe, cygwin1.dll, cygreadline7.dll, cygncursesw-10.dll.
> 
> [calc-2.12.7.2.zip](https://github.com/lcn2/calc/files/10152163/calc-2.12.7.2.zip)

Hello @Leoongithub,

We ask that you edit the _compiling with Cygwin_ section of `README.WINDOWS` in provide a Cygwin user instructions on how to compile calc as your did.

We suggest that you mention the version of Cygwin that you used, as well as a link to where a Windows can obtain it.  We suggest that you indicate the version of Windows that you used as well.  And of course. the command lines you used.  Try to be _tutorial_ as you can, giving a Windows user a recipe for how to compile calc.

We will, in turn, add your zip file to the distribution and reference the _compiling with Cygwin_ section of `README.WINDOWS` that you write.

BTW: We are working on calc version 2.14.2.0 .. so we might want to include a zip file for that release when it comes out.

**Reply by Leoongithub** [reply]
I have created a pr. I'm not native English speaker, so feel free to modify my expression if necessary.
BTW, the zip I offer in this page IS calc-2.14.1.2. I just mistyped the filename.

**Comment by lcn2** [comment]
Hello @Leoongithub .. we expanded on your pr with commit 1b0e6c44629147e77823a9ad36d129066be8cd93.
Comments and corrections welcome.

We have not figured out the best way, if needed, to integrate things like your zip file into the calc repo.

One option is for you to create your own _calc-cygwin_ repo where you could build calc under Cygwin
as needed .. perhaps when new stable version of calc are leased (which is not that often, BTW).
If you did that, then we would be happy to reference at such a repo of yours in the `README.WINDOWS` file.

We are open to other suggestions.

**Comment by lcn2** [comment]
We recommend using Cygwin [Msys2](https://www.msys2.org) to build calc under Windows.  The file `README.WINDOWS` has been updated with the most recent information.

We will close this discussion with our thanks.
